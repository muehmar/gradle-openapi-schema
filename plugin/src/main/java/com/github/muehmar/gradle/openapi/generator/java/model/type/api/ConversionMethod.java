package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class ConversionMethod {
  private final FactoryMethodConversion factoryMethodConversion;
  private final InstanceMethodConversion instanceMethodConversion;

  private ConversionMethod(
      FactoryMethodConversion factoryMethodConversion,
      InstanceMethodConversion instanceMethodConversion) {
    this.factoryMethodConversion = factoryMethodConversion;
    this.instanceMethodConversion = instanceMethodConversion;
  }

  public static ConversionMethod ofString(String conversionMethod) {
    return FactoryMethodConversion.fromString(conversionMethod)
        .map(ConversionMethod::ofFactoryMethod)
        .orElseGet(
            () ->
                ConversionMethod.ofInstanceMethod(
                    InstanceMethodConversion.ofString(conversionMethod)));
  }

  static ConversionMethod ofFactoryMethod(FactoryMethodConversion factoryMethodConversion) {
    return new ConversionMethod(factoryMethodConversion, null);
  }

  static ConversionMethod ofInstanceMethod(InstanceMethodConversion instanceMethodConversion) {
    return new ConversionMethod(null, instanceMethodConversion);
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    if (factoryMethodConversion != null) {
      return onFactoryMethod.apply(factoryMethodConversion);
    } else {
      return onInstanceMethod.apply(instanceMethodConversion);
    }
  }
}
