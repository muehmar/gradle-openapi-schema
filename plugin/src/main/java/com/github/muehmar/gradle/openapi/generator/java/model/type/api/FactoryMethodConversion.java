package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.Optional;
import lombok.Value;

@Value
public class FactoryMethodConversion {
  QualifiedClassName className;
  Name methodName;

  public static Optional<FactoryMethodConversion> fromString(
      QualifiedClassName className, String factoryMethodConversion) {
    final String[] parts = factoryMethodConversion.split("#");
    if (parts.length != 2 && factoryMethodConversion.contains("#")) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
    if (parts.length < 2) {
      return Optional.empty();
    }

    try {
      final QualifiedClassName parsedClassName = QualifiedClassName.ofQualifiedClassName(parts[0]);
      final Name methodName = Name.ofString(parts[1]);
      if (not(parsedClassName.getPackageName().isPresent())
          && parsedClassName.getClassName().equals(className.getClassName())) {
        return Optional.of(new FactoryMethodConversion(className, methodName));
      }
      return Optional.of(new FactoryMethodConversion(parsedClassName, methodName));
    } catch (Exception e) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
  }

  private static OpenApiGeneratorException createInvalidFormatException(
      String factoryMethodConversion) {
    return new OpenApiGeneratorException(
        "Invalid configuration for type conversion '%s'. Expected format for static factory methods: 'package.name.ClassName#methodName'.",
        factoryMethodConversion);
  }
}
