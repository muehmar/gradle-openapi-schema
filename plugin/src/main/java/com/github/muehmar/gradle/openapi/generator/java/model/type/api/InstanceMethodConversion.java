package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.Value;

@Value
public class InstanceMethodConversion {
  Name methodName;

  public static InstanceMethodConversion ofString(String methodName) {
    return new InstanceMethodConversion(Name.ofString(methodName));
  }
}
