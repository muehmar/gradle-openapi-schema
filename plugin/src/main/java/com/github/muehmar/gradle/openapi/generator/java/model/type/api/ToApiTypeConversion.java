package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.Value;

@Value
public class ToApiTypeConversion {
  ConversionMethod conversionMethod;

  public ToApiTypeConversion(ConversionMethod conversionMethod) {
    this.conversionMethod = conversionMethod;
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    return conversionMethod.fold(onFactoryMethod, onInstanceMethod);
  }
}
