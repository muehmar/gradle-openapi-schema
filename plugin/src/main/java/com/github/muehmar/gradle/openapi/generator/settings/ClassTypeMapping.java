package com.github.muehmar.gradle.openapi.generator.settings;

import io.github.muehmar.pojobuilder.annotations.Nullable;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.Optional;
import lombok.AllArgsConstructor;
import lombok.Value;

@Value
@AllArgsConstructor
@PojoBuilder
public class ClassTypeMapping implements Serializable {
  String fromClass;
  String toClass;
  @Nullable TypeConversion typeConversion;
  boolean disableMissingConversionWarning;

  public ClassTypeMapping(
      String fromClass, String toClass, Optional<TypeConversion> typeConversion) {
    this.fromClass = fromClass;
    this.toClass = toClass;
    this.typeConversion = typeConversion.orElse(null);
    this.disableMissingConversionWarning = false;
  }

  public Optional<TypeConversion> getTypeConversion() {
    return Optional.ofNullable(typeConversion);
  }
}
