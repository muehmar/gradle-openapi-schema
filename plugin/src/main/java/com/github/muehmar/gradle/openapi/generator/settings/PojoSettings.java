package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
