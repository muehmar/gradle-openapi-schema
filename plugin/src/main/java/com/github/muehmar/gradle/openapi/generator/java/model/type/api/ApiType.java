package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.ParameterizedApiClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeConversion;
import lombok.Value;

@Value
public class ApiType {
  QualifiedClassName className;
  ParameterizedApiClassName parameterizedClassName;
  ToApiTypeConversion toApiTypeConversion;
  FromApiTypeConversion fromApiTypeConversion;

  public static ApiType fromConversion(
      QualifiedClassName className, TypeConversion conversion, PList<JavaType> generics) {
    final ParameterizedApiClassName parameterizedClassName =
        ParameterizedApiClassName.ofClassNameAndGenerics(className, generics);
    final ToApiTypeConversion toApiTypeConversion =
        new ToApiTypeConversion(ConversionMethod.ofString(className, conversion.getToCustomType()));
    final FromApiTypeConversion fromApiTypeConversion =
        new FromApiTypeConversion(
            ConversionMethod.ofString(className, conversion.getFromCustomType()));
    return new ApiType(
        className, parameterizedClassName, toApiTypeConversion, fromApiTypeConversion);
  }
}
