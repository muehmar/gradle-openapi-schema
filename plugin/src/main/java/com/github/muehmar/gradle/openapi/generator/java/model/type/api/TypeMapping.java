package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMapping;
import com.github.muehmar.gradle.openapi.generator.settings.FormatTypeMapping;
import java.util.Optional;
import lombok.Value;

@Value
public class TypeMapping {
  QualifiedClassName className;
  Optional<ApiType> apiType;

  public static TypeMapping fromClassName(QualifiedClassName className) {
    return new TypeMapping(className, Optional.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName, PList<ClassTypeMapping> classTypeMappings) {
    return fromClassMappings(originalClassName, classTypeMappings, PList.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName,
      PList<ClassTypeMapping> classTypeMappings,
      PList<JavaType> generics) {
    return classTypeMappings
        .filter(
            classMapping ->
                classMapping.getFromClass().equals(originalClassName.getClassName().asString()))
        .headOption()
        .map(classMapping -> fromClassMapping(originalClassName, classMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromClassMapping(
      QualifiedClassName originalClassName,
      ClassTypeMapping classTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.fromClassTypeMapping(classTypeMapping);

    return classTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings) {
    return fromFormatMappings(originalClassName, formatString, formatTypeMappings, PList.empty());
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings,
      PList<JavaType> generics) {
    return formatTypeMappings
        .filter(formatMapping -> formatMapping.getFormatType().equals(formatString))
        .headOption()
        .map(formatMapping -> fromFormatMapping(originalClassName, formatMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromFormatMapping(
      QualifiedClassName originalClassName,
      FormatTypeMapping formatTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.ofQualifiedClassName(formatTypeMapping.getClassType());

    return formatTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public TypeMapping or(TypeMapping other, QualifiedClassName originalClassName) {
    if (this.apiType.isPresent()) {
      return this;
    } else if (other.getApiType().isPresent()) {
      return other;
    } else if (this.className.equals(originalClassName)) {
      return other;
    } else {
      return this;
    }
  }
}
