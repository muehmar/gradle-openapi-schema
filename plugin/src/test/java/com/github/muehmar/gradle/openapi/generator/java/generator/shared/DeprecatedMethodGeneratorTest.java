package com.github.muehmar.gradle.openapi.generator.java.generator.shared;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.DeprecatedMethodGenerator.deprecatedJavaDocAndAnnotationForValidationMethod;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static org.junit.jupiter.api.Assertions.assertEquals;

import au.com.origin.snapshots.Expect;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.ValidationMethods;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class DeprecatedMethodGeneratorTest {
  private Expect expect;

  @Test
  void generate_when_disabledDeprecationAnnotation_then_noOutput() {
    final Generator<Object, PojoSettings> generator =
        deprecatedJavaDocAndAnnotationForValidationMethod();

    final Writer writer =
        generator.generate(
            new Object(), TestPojoSettings.defaultTestSettings(), Writer.javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  void generate_when_enabledDeprecationAnnotation_then_matchSnapshot() {
    final Generator<Object, PojoSettings> generator =
        deprecatedJavaDocAndAnnotationForValidationMethod();

    final Writer writer =
        generator.generate(
            new Object(),
            TestPojoSettings.defaultTestSettings()
                .withValidationMethods(new ValidationMethods(JavaModifier.PUBLIC, "Raw", true)),
            Writer.javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
