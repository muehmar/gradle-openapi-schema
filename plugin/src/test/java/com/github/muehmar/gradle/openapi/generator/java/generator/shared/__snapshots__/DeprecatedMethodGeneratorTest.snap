com.github.muehmar.gradle.openapi.generator.java.generator.shared.DeprecatedMethodGeneratorTest.generate_when_enabledDeprecationAnnotation_then_matchSnapshot=[
.
.
/**
 * @deprecated This method is intended to be used only by frameworks for validation
 * or serialisation.
 */
@Deprecated
]